Improve property parsing

aydany opened the pull request:
This set introduces two improvements:

1. The properties cache in the parser could store multiple copies of the same property
This situation could occur because the last property parsed was added to the cache and returned but the buffer pointer and property index were not incremented. This is now fixed: every parsed property can exist in the cache just once and there is no duplicate property parsing.

2. String values (ANSI and UNICODE) that belong to the last property in an event record are not guaranteed to be null-terminated. The code assumes that every string is null-terminated, which causes strlen and wcslen to potentially access invalid memory looking for a null terminator. If the memory access succeeds, the parser will throw an exception as the length of the string property will be too large. These error conditions were exposed when processing events from http.sys. I modeled the fix after the way Microsoft.Diagnostics.Tracing.TraceEventRawReaders deals with strings. All unit tests still pass and the events from http.sys are now processed without errors.

swannman commented:
@aydany thanks for the pull request! I'm reviewing now. Also cc @zacbrown 

Here's the TraceEventRawReaders reference: https://github.com/Microsoft/perfview/blob/master/src/TraceEvent/TraceEvent.cs#L3886

swannman commented:
@aydany would you be willing to write a unit test that covers the non-null-terminated case so we have test coverage on this new code?

swannman commented on krabs/krabs/parser.hpp:
```diff
@@ -179,13 +179,15 @@ namespace krabs {
             // The property was found, return it
             if (name == pName) {
+                // advance the index since we've already processed this property
+                ++i;
```
is this line necessary? if we're about to `return propInfo` then `i` will go out of scope so we don't need to advance it?

aydany replied:
Note that `i` is defined as a reference to  lastPropertyIndex_` at the start of the for loop.

swannman replied:
You're right, I missed the ampersand. Thanks!

aydany commented:
@swannman I'll take a look at adding a test tomorrow. Thanks for the review!

zacbrown commented on krabs/krabs/size_provider.hpp:
```diff
@@ -105,18 +108,35 @@ namespace krabs {
             {
-                // size of ansi string is length + 1 (null byte)
-                propertyLength = (ULONG)strlen((const char*)propertyStart) + 1;
+                const char* p = (const char*)propertyStart;
```
stylistically - I'd prefer to see us use `auto` here.

zacbrown commented on krabs/krabs/size_provider.hpp:
```diff
@@ -105,18 +108,35 @@ namespace krabs {
-                // size of unicode string is length + 1 (null byte) * sizeof(wchar)
-                auto chars = (ULONG)wcslen((const wchar_t*)propertyStart) + 1;
-                propertyLength = chars * sizeof(wchar_t);
+                const wchar_t* p = (const wchar_t*)propertyStart;
```
Stylistically - prefer `auto` here.

zacbrown replied:
ditto line below as well

zacbrown commented on krabs/krabs/size_provider.hpp:
```diff
@@ -105,18 +108,35 @@ namespace krabs {
             {
-                // size of ansi string is length + 1 (null byte)
-                propertyLength = (ULONG)strlen((const char*)propertyStart) + 1;
+                const char* p = (const char*)propertyStart;
```
Prefer `auto` here

zacbrown replied:
ditto line below as well

swannman commented:
@aydany it looks like there are a few minor changes necessary before we can merge this in. Are you still willing to make those?

aydany commented:
I apologize for taking way too long to get back to this. I addressed the comment about style and added a unit test to cover the code changes. The test will fail when run against master, but passes here.

Please note that the test covers only the UNICODE string case. To cover the ANSI string case we need a valid ETW event that has an ANSI string as its last property. I happened to encounter the UNICODE case accidentally and I do not have an example that can be used for the ANSI case at this time.
